include/oneapi/dpl/experimental/kt/internal/esimd_radix_sort_dispatchers.h

```diff
@@ -266,33 +243,132 @@ __onesweep(sycl::queue __q, _RngPack&& __pack, ::std::size_t __n)
+    __mem_holder.__group_hist_item_alloc_count(__group_hist_item_count);
+    __mem_holder.__allocate();
+}
+
+template <typename _RngPack, typename _MemHolder>
+auto
+__create_temp_pack(const _MemHolder& __mem_holder, ::std::size_t __n)
+{
+    using _KeyT = typename ::std::decay_t<_RngPack>::_KeyT;
+    using _ValT = typename ::std::decay_t<_RngPack>::_ValT;
+    constexpr bool __has_values = ::std::decay_t<_RngPack>::__has_values;
+
+    auto __keys_tmp_keep = oneapi::dpl::__ranges::__get_sycl_range<sycl::access_mode::read_write, _KeyT*>();
+    auto __keys_tmp_rng = __keys_tmp_keep(__mem_holder.__keys_ptr(), __mem_holder.__keys_ptr() + __n).all_view();
+    if constexpr (__has_values)
+    {
+        auto __vals_tmp_keep = oneapi::dpl::__ranges::__get_sycl_range<sycl::access_mode::read_write, _ValT*>();
+        auto __vals_tmp_rng = __vals_tmp_keep(__mem_holder.__vals_ptr(), __mem_holder.__vals_ptr() + __n).all_view();
+        return __rng_pack(::std::move(__keys_tmp_rng), ::std::move(__vals_tmp_rng));
+    }
+    else
+    {
+        return __rng_pack(::std::move(__keys_tmp_rng));
+    }
+}
+
+template <typename _KernelName, bool __is_ascending, ::std::uint8_t __radix_bits, ::std::uint16_t __data_per_work_item,
+          ::std::uint16_t __work_group_size, bool __in_place, typename _RngPack1, typename _RngPack2>
+sycl::event
+__onesweep(sycl::queue __q, _RngPack1&& __pack, _RngPack2&& __pack_out, ::std::size_t __n)
+{
+    using _KeyT = typename ::std::decay_t<_RngPack1>::_KeyT;
+    using _ValT = typename ::std::decay_t<_RngPack1>::_ValT;
+    constexpr bool __has_values = ::std::decay_t<_RngPack1>::__has_values;
+
+    using _EsimdRadixSortCopyback = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<
+        ::std::conditional_t<__has_values, __esimd_radix_sort_onesweep_copyback_by_key<_KernelName>,
+                             __esimd_radix_sort_onesweep_copyback<_KernelName>>>;
+
+    using _GlobalHistT = ::std::uint32_t;
+
+    constexpr ::std::uint32_t __bit_count = sizeof(_KeyT) * 8;
+    constexpr ::std::uint32_t __stage_count = oneapi::dpl::__internal::__dpl_ceiling_div(__bit_count, __radix_bits);
+
+    // Memory is not going to be allocated for void value type
+    // TODO: make this more explicit to reduce coupling between __onesweep_memory_holder and __rng_pack
+    __onesweep_memory_holder<_GlobalHistT, _KeyT, _ValT> __mem_holder(__q);
+
+    __allocate_temp<__radix_bits, __data_per_work_item, __work_group_size, _RngPack1>(__mem_holder, __n);
+    auto __tmp_pack = __create_temp_pack<_RngPack1>(__mem_holder, __n);
+    auto select_element = [] {
+        if constexpr (__in_place)
+            return 0;
+        else
+            return (__stage_count % 2);
+    };
+
+    auto __virt_pack1 = ::std::get<select_element()>(::std::tuple(__tmp_pack, __pack_out));
+    auto __virt_pack2 = ::std::get<select_element()>(::std::tuple(__pack_out, __tmp_pack));
```

Comment:
This is an interesting way to statically select from one of two variables based on a constant expression. I did some testing on Godbolt and from what I can tell, it seems like these two lines would create 2 copies each of `__tmp_pack` and `__pack_out`: https://godbolt.org/z/nT1sMfMdv.

I'm not sure if these copies are important to avoid, but I thought it was an interesting experiment to test out.

Reply:
See the response to Matt above, 
I'll experiment a bit in godbolt, and just make a constexpr selection function using const refs.

Reply:
I've simplified it away from `tuple` and confirmed this works without copies https://godbolt.org/z/Pe45WeYKT